datafusion/core/src/datasource/listing/path.rs

```diff
@@ -0,0 +1,217 @@
+    ///
+    /// If no scheme is provided, the string will be interpreted as a
+    /// path on the local filesystem, using the operating system's
+    /// standard path delimiter - i.e. `\` on Windows, `/` on Unix.
+    ///
+    /// If the path contains any of `'?', '*', '['`, it will be considered
+    /// a glob expression and resolved as described in the section below.
+    ///
+    /// Otherwise, the path will be resolved to an absolute path, returning
+    /// an error if it does not exist, and converted to a [file URI]
+    ///
+    /// If you wish to specify a path that does not exist on the local
+    /// machine you must provide it as a fully-qualified [file URI]
+    /// e.g. `file:///myfile.txt`
+    ///
+    /// ## Glob Paths
+    ///
+    /// If no scheme is provided, and the path contains a glob expression, it will
+    /// be resolved as follows.
+    ///
+    /// The string up to the first path segment containing a glob expression will be extracted,
+    /// and resolved in the same manner as a normal scheme-less path. That is, resolved to
+    /// an absolute path on the local filesystem, returning an error if it does not exist,
+    /// and converted to a [file URI]
+    ///
+    /// The remaining string will be interpreted as a [`glob::Pattern`] and used as a
+    /// filter when listing files from object storage
+    ///
+    /// [file URI]: https://en.wikipedia.org/wiki/File_URI_scheme
+    pub fn parse(s: impl AsRef<str>) -> Result<Self> {
+        let s = s.as_ref();
+        Ok(match Url::parse(s) {
+            Ok(url) => Self { url, glob: None },
+            Err(url::ParseError::RelativeUrlWithoutBase) => {
+                let (prefix, glob) = match split_glob_expression(s) {
+                    Some((prefix, glob)) => {
+                        let glob = Pattern::new(glob)
+                            .map_err(|e| DataFusionError::External(Box::new(e)))?;
+                        (prefix, Some(glob))
+                    }
+                    None => (s, None),
+                };
+
+                let path = std::path::Path::new(prefix).canonicalize()?;
+                let url = match path.is_file() {
+                    true => Url::from_file_path(path).unwrap(),
+                    false => Url::from_directory_path(path).unwrap(),
+                };
+
+                Self { url, glob }
+            }
+            Err(e) => return Err(DataFusionError::External(Box::new(e))),
+        })
+    }
+
+    /// Returns the URL scheme
+    pub fn scheme(&self) -> &str {
+        self.url.scheme()
+    }
+
+    /// Returns the path as expected by [`ObjectStore`]
+    ///
+    /// In particular for file scheme URLs, this has a leading `/`
+    /// and describes an absolute path on the local filesystem
+    ///
+    /// For other URLs, this also contains the host component
+    /// and lacks a leading `/`
+    ///
+    /// TODO: Handle paths consistently (#2489)
+    fn prefix(&self) -> &str {
```

Comment:
As part of #2489 I hope to bring some consistency to what the paths passed to the ObjectStore actually are. It's currently a bit confused